consensus/enclave/impl/src/lib.rs

```diff
@@ -1127,6 +1159,97 @@ mod tests {
         );
     }
 
+    #[test_with_logger]
+    fn test_client_tx_propose_works(logger: Logger) {
+        let mut rng = Hc128Rng::from_seed([1u8; 32]);
+
+        let block_version = BlockVersion::MAX;
+        let responder_id = Default::default();
+        let fee_map = FeeMap::default();
+
+        let enclave = SgxConsensusEnclave::new(logger.clone());
+        let blockchain_config = BlockchainConfig {
+            block_version,
+            fee_map: fee_map.clone(),
+            ..Default::default()
+        };
+        enclave
+            .enclave_init(&responder_id, &responder_id, &None, blockchain_config)
+            .unwrap();
+
+        // First, create an encrypted connection from a client
+        let initiator = Start::new(responder_id.to_string());
+        let init_input = ClientInitiate::<X25519, Aes256Gcm, Sha512>::default();
+        let (initiator, auth_request_output) = initiator.try_next(&mut rng, init_input).unwrap();
+
+        let auth_request_msg = AuthMessage::from(auth_request_output);
+        let (auth_response_msg, _client_session) =
+            enclave.client_accept(auth_request_msg.into()).unwrap();
```

Comment:
@jcape @nick-mobilecoin I tried to write a unit test per Eran's request that exercises the `client_tx_propose_v2` pathway. However, doing this requires creating an attested connection within the unit test, since that function takes an `attest::Message` argument.

I think I got that part right, but now I'm failing ike so:

```
running 1 test
test tests::test_client_tx_propose_works ... FAILED

failures:

---- tests::test_client_tx_propose_works stdout ----
thread 'tests::test_client_tx_propose_works' panicked at 'called `Result::unwrap()` on an `Err` value: Attest(NoReportAvailable)', consensus/enclave/impl/src/lib.rs:1193:103
```

I remember that we had some difficulties writing tests like this in the past, and may have decided not to write any of them because of this, since it will also all be influened by `SGX_MODE=HW` vs. `SGX_MODE=SW`, and I want to make sure I approach this in the best way.

* The only way to get the enclave to have a report is for it to accept a report via the reportable enclave API.
* I have to do the whole quoting enclave / RA client thing
* It seems like I can use the `mc-sgx-report-cache-untrusted` code in the unit test most likely, and then I think I only need to supply an RA client, and that might avoid some boilerplate.

The thing I'm worried about is that this unit test does not actually run in an enclave, so the `attest-ake` stuff may not work or may fail when it tries to create `Report::new()`, are those e-report instructions even going to work?

I remember that you said in the past that in simulation mode, the quoting enclave will accept anything. But I'm not sure if I can assume that this will run in simulation mode.

The other idea that I had was to go to `mc-crypto-attest-ake` and add a "test-only" feature gating a function that lets me inject a dummy `VerificationReport`. It seems like this is all going to get changed for DCAP anyways so maybe I shouldn't waste too much time on this.

LMK if you guys have thoughts.

I guess another approach would be to move the interesting bits of `client_tx_propose_v2` to a helper function that is called after the decryption happens, and then test that instead.

Reply:
I doubt report will work outside of on enclave it will likely pull in SGX libs, which will then rely on enclave specific functions.

As long as there isn't a concern with info leakage I would vote for moving the encryption/decryption out to a higher level function that is excersized with an integration test, where the unit test focuses on the edge case and specifics of the behavior while the integration tests one happy path and one failure path (unless there is different handling for different failures).

Since we've been talking about testing levels, one of the criteria that is often used to define unit tests is "fast".  While there is no good criteria for _fast_, I usually think of single digit ms, or less, with a few outlier tests hitting 100ms or so.

This takes 4 seconds on my machine.  
```rust
        let mut ledger = create_ledger();
        let sender = AccountKey::random(&mut rng);
        let recipient = AccountKey::random(&mut rng);
        let n_blocks = 3;
        initialize_ledger(block_version, &mut ledger, n_blocks, &sender, &mut rng);
```

Reply:
:+1:

the stuff where we have to create an LMDB database to get test data is some junk left behind by engineers who no longer work here.

Reply:
I'm going to try to unit test this in `mc-consensus-enclave` instead. I don't feel super good about splitting up `client_propose_tx` into two functions just so that there is a private function that we can test. I think testing private functions has little value, I'd rather just do whatever I can to test the public API.

Reply:
gonna mark this resolved in favor of this approach: https://github.com/mobilecoinfoundation/mobilecoin/pull/2670#discussion_r997458095